File: 7.1_Brackets.js
// Determine whether a given string of parentheses is properly nested
function Brackets(S) {
    // split string into array
    var S = S.split('');
    // create object to match lefts and rights
    var match = { "]": "[", "}": "{", ")": "(" };
    // create stack to track bracket matches
    var stack = [];
    // loop through S
    for (var i = 0; i < S.length; i++) {
        // test for leftness (as opposed to rightness)
        if (match[S[i]] == undefined) {
            // if left, add to stack
            stack.push(S[i]);
            // if right, make sure stack already has an element
        }
        else if (stack.length != 0) {
            // remove an element from stack
            var POP = stack.pop();
            // removed element should equal S[i]
            if (match[S[i]] != POP) {
                // if not, return zero
                return (0);
            }
        }
    }
    // if stack.length==0 after loop finishes, the lefts and right 'balance'
    if (stack.length == 0) {
        return (1);
        // otherwise, they do not
    }
    else {
        return (0);
    }
}
// Bash Testing:
var args = process.argv[2];
// args = args.replace(/^\[|\]$/g, "").split(",").map(function (item) { return (parseInt(item)); });
var time1 = Date.now();
var result = Brackets(args);
var time2 = Date.now();
console.log("result: " + result + " ; milliseconds: " + (time2 - time1));
// Codility Testing:
/// https://app.codility.com/demo/results/trainingJ3BU8D-N8G/
/// Correctness: 100% 
/// Performance: 80%
/// Difficulty: Painless

File: 7.1_Brackets.py
# Determine whether a given string of parentheses is properly nested
# Bash Testing:

def Brackets(S):
	# split string into array
	S = list(S)
	# create object to match lefts and rights
	match = {"]":"[","}":"{",")":"("}
	lefts = ["{","[","("]
	# create stack to track bracket matches
	stack = []
	# loop through S
	for i in range(0,len(S)):
		# test for leftness
		if S[i] in lefts:
			# if left, add to stack
			stack.append(S[i])
		# if right, make sure stack already has an element
		elif len(stack)!=0:
			# remove an element from stack
			POP = stack.pop()
			# removed element should equal S[i]
			if match[S[i]]!=POP:
				# if not, return zero
				return(0)
	# if stack.length==0 after loop finishes, the lefts and right 'balance'
	if len(stack)==0:
		return(1)
	# otherwise, they do not
	else:
		return(0)

# Bash Testing:
import getopt, sys, re, datetime
args = list(getopt.getopt(sys.argv,"ho:v"))[1][1]
#args = re.sub("^\[|\]$","",args).split(",")
#args = [int(x) for x in args]
time1 = datetime.datetime.now()
result = Brackets(args)
time2 = datetime.datetime.now()
timeChange = time2-time1
print(["result: "]+[result]+[" ; milliseconds: "]+[timeChange.total_seconds()*1000])
# Codility Testing:
## https://app.codility.com/demo/results/trainingB73C4A-TQM/
## Correctness: 100%
## Performance: 80%
## Difficulty: Painless


